### PR TITLE
Moved HTML entities from DATA section to separate resources file.

### PR DESCRIPTION
**Pull request by rage311**
### Summary
Move HTML entities from Mojo::Util DATA section to separate resources file (#1409)

### Motivation
These changes were requested.

### References
#1409 

The use of `require` helps to circumvent some circular dependencies that occur otherwise.  I'm not sure if this is exactly what was envisioned, and I'm certainly open to input.

### CODE REVIEW

**Comment by kraih** [comment]
I really don't like the inline requires. Perhaps the code would be cleaner if we just didn't use `Mojo::File` in this one case?

**Comment by jberger** [comment]
> I really don't like the inline requires. Perhaps the code would be cleaner if we just didn't use `Mojo::File` in this one case?

I can agree with that. I'd suggest putting a comment in like "don't use Mojo::File here because of circular dependencies" or somethinf

**Comment by Grinnz** [comment]
sibling makes it a little ugly to avoid it, can be cleaned up a bit but: `File::Spec->catfile(File::Basename::dirname(Cwd::realpath(__FILE__)), 'resources', 'html_entities.txt')`

**Comment by rage311** [comment]
I reverted all the `require`s and used @Grinnz solution for a `sibling` equivalent.

File: lib/Mojo/Util.pm [hunk 1]
```diff
@@ -37,11 +38,23 @@ use constant {
 # To generate a new HTML entity table run this command
 # perl examples/entities.pl
 my %ENTITIES;
-for my $line (split "\n", join('', <DATA>)) {
-  next unless $line =~ /^(\S+)\s+U\+(\S+)(?:\s+U\+(\S+))?/;
-  $ENTITIES{$1} = defined $3 ? (chr(hex $2) . chr(hex $3)) : chr(hex $2);
+{
+  # Don't use Mojo::File here due to circular dependencies
+  my $html_path
+    = File::Spec->catfile(File::Basename::dirname(Cwd::realpath(__FILE__)),
```
Comment (kraih) [inline lib/Mojo/Util.pm]: Why `Cwd::realpath` here?

File: lib/Mojo/Util.pm [hunk 2]
```diff
@@ -37,11 +38,23 @@ use constant {
...
+    'resources', 'html_entities.txt');
+
+  local $/ = undef;
```
Comment (kraih) [inline lib/Mojo/Util.pm]: Why do this on an extra line instead of using the `do` idiom?

File: lib/Mojo/Util.pm (hunk 1 above)
Comment (Grinnz) [reply lib/Mojo/Util.pm]: It handles the edge case where Mojo::Util is loaded via a symlink to still reference the correct location

File: lib/Mojo/Util.pm (hunk 2 above)
Comment (rage311) [reply lib/Mojo/Util.pm]: Like this?
`my $file_content = do { local $/; <$fh> };`

I wasn't familiar with that pattern, I had to Google it.

File: lib/Mojo/Util.pm (hunk 1 above)
Comment (kraih) [reply lib/Mojo/Util.pm]: How could that ever happen? Until your recent addition of `curfile` we did not use `Cwd::realpath` for any of this and there were never any issues.

File: lib/Mojo/Util.pm (hunk 2 above)
Comment (kraih) [reply lib/Mojo/Util.pm]: Yes, that's the standard way of using `local $/` actually.

File: lib/Mojo/Util.pm (hunk 1 above)
Comment (rage311) [reply lib/Mojo/Util.pm]: It does seem to work the same if I remove `Cwd::realpath`.

File: lib/Mojo/Util.pm [hunk 3]
```diff
@@ -7,6 +7,7 @@ use Digest::MD5 qw(md5 md5_hex);
 use Digest::SHA qw(hmac_sha1_hex sha1 sha1_hex);
 use Encode 'find_encoding';
 use Exporter 'import';
+use File::Basename;
```
Comment (kraih) [inline lib/Mojo/Util.pm]: I'm almost certain there are default exports polluting the namespace now.

Comment (karenetheridge) [reply lib/Mojo/Util.pm]: You can use Test::CleanNamespaces in an xt/author test to know for sure.

Comment (Grinnz) [reply lib/Mojo/Util.pm]: Yes File::Basename has default exports, can be suppressed with `use File::Basename ()`

**Comment by rage311** [comment]
Imported `File::Basename 'dirname'` only.
Changed to `do { ... }` pattern for setting the record separator and slurping the file.
Removed `Cwd::realpath`.

File: lib/Mojo/Util.pm [hunk 4]
```diff
@@ -37,11 +38,21 @@ use constant {
...
-for my $line (split "\n", join('', <DATA>)) {
-  next unless $line =~ /^(\S+)\s+U\+(\S+)(?:\s+U\+(\S+))?/;
-  $ENTITIES{$1} = defined $3 ? (chr(hex $2) . chr(hex $3)) : chr(hex $2);
+{
+  # Don't use Mojo::File here due to circular dependencies
+  my $html_path = File::Spec->catfile(dirname(__FILE__),
+    'resources', 'html_entities.txt');
+
+  open my $html_fh, '<', $html_path
+    or croak 'Unable to open html entities file';
```
Comment (kraih) [inline lib/Mojo/Util.pm]: I don't like hiding the actual error, especially when it would be such an unexpected exception.

File: lib/Mojo/Util.pm [hunk 5]
```diff
@@ -37,11 +38,21 @@ use constant {
...
+  my $entity_lines = do { local $/; <$html_fh> };
+  close $html_fh;
```
Comment (kraih) [inline lib/Mojo/Util.pm]: Do we gain anything by not just letting it go out of scope here?

Comment (rage311) [reply lib/Mojo/Util.pm]: A bit of explicitness, I suppose, but I don't feel strongly about it either way.

**Comment by kraih** [comment]
Thanks, merged.

**Pull request merged**